### PR TITLE
Add PathParam Warning Diagnostic

### PR DESCRIPTION
**Pull request by TheAvi123**
Issue #245

### CODE REVIEW

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java [hunk 1]
```diff
@@ -32,6 +37,78 @@ public void completeDiagnostic(Diagnostic diagnostic) {
...
+    				List<String> endpointPathVars = findAndProcessEndpointURI(type);
+    				if (endpointPathVars == null) {
+    					continue;
+    				}
+    				IMethod[] typeMethods = type.getMethods();
+    				for (IMethod method : typeMethods) {
+    					ILocalVariable[] methodParams = method.getParameters();
+    					for (ILocalVariable param : methodParams) {
+    						IAnnotation[] paramAnnotations = param.getAnnotations();
+    	    				for (IAnnotation annotation : paramAnnotations) {
```
Comment (dalidia) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java:54]: Insignificant issue but formatting is off in this part and its corresponding closing statement

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java [hunk 2]
```diff
@@ -32,6 +37,78 @@ public void completeDiagnostic(Diagnostic diagnostic) {
...
+    		}
+    	}
+    }
+    
+    private Diagnostic createPathParamWarningDiagnostic(IJavaElement annotation, IOpenable unit) throws JavaModelException {
+		ISourceRange nameRange = JDTUtils.getNameRange(annotation);
+		Range range = JDTUtils.toRange(unit, nameRange.getOffset(), nameRange.getLength());
+		Diagnostic diagnostic = new Diagnostic(range, WebSocketConstants.PATHPARAM_VALUE_WARN_MSG);
+		diagnostic.setSource(WebSocketConstants.DIAGNOSTIC_SOURCE);
+        diagnostic.setSeverity(WebSocketConstants.WARNING);
```
Comment (dalidia) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java:82]: Formatting is a bit off. Make sure you're using spaces

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java [hunk 3]
```diff
@@ -32,6 +37,78 @@ public void completeDiagnostic(Diagnostic diagnostic) {
...
+        diagnostic.setCode("ChangePathParamValue");
+		return diagnostic;
+    }
+    
+    private List<String> findAndProcessEndpointURI(IType type) throws JavaModelException {
+    	String endpointURI = null;
+    	IAnnotation[] typeAnnotations = type.getAnnotations();
+		for (IAnnotation annotation : typeAnnotations) {
```
Comment (dalidia) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java:90]: Formatting is a bit off here

**Review by dalidia — CHANGES_REQUESTED** [review]
It looks good overall. I left some minor comments

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.test/src/main/java/org/eclipse/lsp4jakarta/jdt/websocket/JakartaWebSocketTest.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
+/*******************************************************************************
+* Copyright (c) 2021 IBM Corporation and others.
```
Comment (turkeylurkey) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.test/src/main/java/org/eclipse/lsp4jakarta/jdt/websocket/JakartaWebSocketTest.java:2]: Year 2022 please.

**Review by turkeylurkey — CHANGES_REQUESTED** [review]

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java [hunk 5]
```diff
@@ -32,6 +37,78 @@ public void completeDiagnostic(Diagnostic diagnostic) {
...
+			if (annotation.getElementName().equals(WebSocketConstants.WEBSOCKET_SERVER_ANNOTATION) || 
+				annotation.getElementName().equals(WebSocketConstants.WEBSOCKET_CLIENT_ANNOTATION)) {
+				IMemberValuePair[] valuePairs = annotation.getMemberValuePairs();
+				for (IMemberValuePair pair : valuePairs) {
+					if (pair.getMemberName().equals(WebSocketConstants.ANNOTATION_VALUE)) {
+						endpointURI = (String) pair.getValue();
```
[the remark this answers is not in the document]
Comment (TheAvi123) [reply jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java:96]: Fixed

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java [hunk 6]
```diff
@@ -32,6 +37,78 @@ public void completeDiagnostic(Diagnostic diagnostic) {
...
+    				for (IMethod method : typeMethods) {
+    					ILocalVariable[] methodParams = method.getParameters();
+    					for (ILocalVariable param : methodParams) {
+    						IAnnotation[] paramAnnotations = param.getAnnotations();
+    	    				for (IAnnotation annotation : paramAnnotations) {
+    	    					if (annotation.getElementName() == "PathParam") {
+    	    						IMemberValuePair[] valuePairs = annotation.getMemberValuePairs();
+    	    						for (IMemberValuePair pair : valuePairs) {
+    	    							if (pair.getMemberName().equals(WebSocketConstants.ANNOTATION_VALUE)) {
+    	    								String pathValue = (String) pair.getValue();
```
[the remark this answers is not in the document]
Comment (TheAvi123) [reply jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java:59]: Fixed

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java (hunk 2 above)
Comment (TheAvi123) [reply jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java:88]: Resolved

**Review by turkeylurkey — APPROVED** [review]

**Review by kathrynkodama — APPROVED** [review]

**Pull request merged**